### PR TITLE
Boetseren

### PR DESCRIPTION
**Pull request by geralddejong**

### CODE REVIEW

File: client/src/fabric/features.ts [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+    setFactor: (factor: number) => void
+    factor$: BehaviorSubject<number>
+    defaultValue: number
+}
+
+export function applyFeatureToInstance(instance: FabricInstance):  (feature: IFeature) => void {
+    return feature => {
+        const intervalRole = feature.name.intervalRole
+        if (intervalRole === undefined) {
+            throw new Error()
```
Comment (mvanderh) [inline client/src/fabric/features.ts]: Throw in an error message for the poor developers

File: client/src/fabric/features.ts [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+        engine.setPhysicsFeature(globalFeature, factor)
+    }
+}
+
+export interface IFeatureStorage {
+    getFeature: (label: string, defaultValue: number) => number
+    setFeature: (label: string, factor: number) => void
+}
+
+export function enumToFeatureArray(enumObject: object, physics: boolean, storage: IFeatureStorage): IFeature[] {
```
Comment (mvanderh) [inline client/src/fabric/features.ts]:
```suggestion
export function enumToFeatureArray(enumObject: object, isPhysics: boolean, storage: IFeatureStorage): IFeature[] {
```
To be consistent

File: client/src/fabric/features.ts [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+    return Object.keys(enumObject)
+        .filter(key => key.length > 1)
```
Comment (mvanderh) [inline client/src/fabric/features.ts]: Is this to filter out the reverse keys e.g. "1" -> "SomeEnumValue"? If so, this'll break once you add enum value #10

File: client/src/fabric/features.ts [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+    if (feature.physicsFeature !== undefined) {
+        return PhysicsFeature[feature.physicsFeature]
+    }
+    if (feature.intervalRole !== undefined) {
+        return IntervalRole[feature.intervalRole]
+    }
+    return "?"
+}
+
+export interface IFeature {
```
Comment (mvanderh) [inline client/src/fabric/features.ts]: Why isn't this a class? "setFactor()" looks like a method to me

File: client/src/fabric/tensegrity-fabric.ts [hunk 5]
```diff
@@ -243,7 +235,7 @@ export class TensegrityFabric {
                             break
                     }
                 } else {
-                    this.physics.applyLocalFeatures(this.instance)
+                    this.roleFeatures.forEach(applyFeatureToInstance(this.instance))
```
Comment (mvanderh) [inline client/src/fabric/tensegrity-fabric.ts]: Is this gonna be a convention in the project now? It's interesting

File: client/src/view/adjust-panel.tsx [hunk 6]
```diff
@@ -82,15 +65,16 @@ export function AdjustPanel({fabric, setDisplacementSelection}: {
 
     const IntervalGroupToggle = () => {
         const setAdjustBars = (bars: boolean) => {
```
Comment (mvanderh) [inline client/src/view/adjust-panel.tsx]:
```suggestion
        const setAdjustBars = (barsMode: boolean) => {
```

File: client/src/view/adjust-panel.tsx [hunk 7]
```diff
@@ -194,6 +167,59 @@ export function AdjustPanel({fabric, setDisplacementSelection}: {
...
+        borderRadius: 12,
+    }
+
+    const changeNuanceTo = (nuanceValue: number) => {
+        setNuance(nuanceValue)
+        setDisplacement(displacementFromNuance(nuanceValue))
+        setFabricSlackLimits(adjustBars, nuanceValue)
+    }
+
+    const onChange = (changedPercent: number[]) => {
```
Comment (mvanderh) [inline client/src/view/adjust-panel.tsx]:
```suggestion
    const onChange = ([percent]: number[]) => {
```
I figure you can deconstruct here

File: client/src/view/adjust-panel.tsx [hunk 8]
```diff
@@ -194,6 +167,59 @@ export function AdjustPanel({fabric, setDisplacementSelection}: {
...
+        const percent = changedPercent[0]
+        changeNuanceTo(percent / 100)
+    }
+
+    return (
+        <Slider rootStyle={sliderStyle} reversed={true} vertical={true}
+                step={-0.01} domain={DOMAIN} values={VALUES}
+                onChange={onChange}>
+            <Rail>{({getRailProps}) => <SliderRail getRailProps={getRailProps}/>}</Rail>
```
Comment (mvanderh) [inline client/src/view/adjust-panel.tsx]: Whoa how does this work? A function inside of a JSX element?

File: client/src/fabric/features.ts (hunk 1 above)
Comment (geralddejong) [reply client/src/fabric/features.ts]: Yes.

File: client/src/fabric/features.ts (hunk 2 above)
Comment (geralddejong) [reply client/src/fabric/features.ts]: Yes

File: client/src/fabric/tensegrity-fabric.ts (hunk 5 above)
Comment (geralddejong) [reply client/src/fabric/tensegrity-fabric.ts]: Things are getting clearer.

File: client/src/view/adjust-panel.tsx (hunk 6 above)
Comment (geralddejong) [reply client/src/view/adjust-panel.tsx]: Yes

File: client/src/view/adjust-panel.tsx (hunk 8 above)
Comment (geralddejong) [reply client/src/view/adjust-panel.tsx]: It was the easiest slider for me to integrate.

File: client/src/fabric/features.ts (hunk 4 above)
Comment (geralddejong) [reply client/src/fabric/features.ts]: Could be.

File: client/src/fabric/features.ts (hunk 3 above)
Comment (geralddejong) [reply client/src/fabric/features.ts]: Yes.  Fixed that.

**Pull request merged**